doc/adr/0012-modules-structure.md

```diff
@@ -0,0 +1,105 @@
+# 12. Modules' structure
+
+Date: 2022-12-28
+
+## Status
+
+Accepted
+
+## Context
+
+The existing structure of tessellation modules was mixed, hence hard to maintain.
+
+- `shared` contains both base classes and extensions, and tessellation schema,
+- `keytool` contains key-pair generator and exposes key-related operations,
+- `wallet` contains operations related to transactions,
+- `kernel` contains part of state channels (hylomorphism's idea),
+- `sdk` contains all the things shared between Global L0 and DAG L1, but not the $DAG itself,
+- `dag-shared` contains all the things shared between Global L0 and DAG L1 related to $DAG,
+- `core` is the Global L0 module (but also contains state channels that must be extracted to be reused in other modules),
+- `dag-l1` is the DAG L1 module.
+
+```mermaid
+graph TD
+    A[Shared]
+    B[Keytool]
+    C[Kernel]
+    D[Wallet]
+    E[SDK]
+    F[DAG Shared]
+    G[Core]
+    H[DAG L1]
+    
+    B --> A
+    C --> A
+    D --> B
+    D -.-> A
+    E --> C
+    E -.-> B
+    E -.-> A
+    E --> F
+    F --> C
+    F -.-> B
+    F -.-> A
+    G --> E
+    G -.-> C
+    G -.-> B
+    G -.-> A
+    G --> F
+    H --> F
+    H --> E
+    H -.-> C
+    H -.-> B
+    H -.-> A
+```
+
+Furthermore, dependencies were partially explicitly declared and partially transient. After some time, it became unmaintainable.
+
+The second reason for restructuring modules is the fact that 3rd parties cannot use the SDK module in the current form. It does not expose all required operations, so internal modules were used directly, leaking the project's internal structure and locking further internal changes.
+
+## Decision
+
+To simplify the structure of modules, we will split `shared` into two separate modules - the first containing scala and libraries extensions, the latter containing the tessellation schema.
+Then, all the things related to node operations would be kept in `node-shared` module (including gossip protocol, consensus, etc.). This leads to emptying the SDK module, so we will be able to implement the SDK module as a proxy between the external world and the internal tessellation structure. SDK module should reexpose all necessary operations to be used by 3rd parties. For example - to create a keypair, global L0 can use the keytool module directly, but 3rd party will use the SDK module that will reexpose keypair creation function. That way, the internal structure of tessellation will be hidden behind the facade.
+
+Modules will be transient - for example: `wallet` does not need to depend on `schema` directly, because it depends on `keytool` that depends on `schema`.
```

Comment:
Wait, I thought we want to avoid transitive dependencies? So If some module is using classes from `shared` it's not enough to depend just on the `schema`, it has to depend on `shared`.